msi/msi.go

```diff
@@ -65,11 +65,35 @@ func (p *provider) getMsiHelper(queryParams map[string]string) (*Msi, error) {
 	}
 	requestUrl.RawQuery = urlQuery.Encode()
 
-	code, body, err := p.httpClient.Get(requestUrl.String(), map[string]string{"Metadata": "true"})
+	code, respHeaders, body, err := p.httpClient.GetWithHeaders(requestUrl.String(), map[string]string{"Metadata": "true"})
 	if err != nil {
 		return &msi, err
 	}
 
+	// Arc uses a challenge response mechanism to get the token
+	// If the response code is 401, Arc will have a header Www-Authenticate: Basic realm=<location of the token>
+	if GetMetadataIdentityURL() != metadataIdentityURL && code == 401 {
+		wwwAuthenticateHeader, exists := respHeaders["Www-Authenticate"]
+		if !exists || len(wwwAuthenticateHeader) == 0 {
+			return &msi, errorhelper.AddStackToError(fmt.Errorf("unable to get msi, metadata service response code %v: Www-Authenticate header missing or empty", code))
+		}
+		tokenLocation := wwwAuthenticateHeader[0]
+		if len(tokenLocation) == 0 {
+			return &msi, errorhelper.AddStackToError(fmt.Errorf("unable to get msi, metadata service response code %v: token location is empty", code))
+		}
+
+		tokenLocation = tokenLocation[len("Basic realm="):]
+		token, err := os.ReadFile(tokenLocation)
+		if err != nil {
```

Comment:
Similar comment to Windows. We need to limit from where the file can be loaded, and put some limitation of the format/name of the file. Otherwise, a malicious host could use this to retrieve any file on the VM.